crates/holochain/src/core/workflow/sys_validation_workflow.rs

```diff
@@ -43,170 +42,410 @@ mod validate_op_tests;
-                                        )?;
-                                    }
-                                }
-                            }
-                        }
-                        WorkflowResult::Ok(summary)
-                    })
-                    .await
+                let fetched = cascade.retrieve_action(h.clone(), Default::default()).await;
+                tracing::trace!(hash = ?h, fetched = ?fetched, "Fetched action for validation");
+                (h, fetched)
             }
             .boxed()
-        },
+        });
+
+    let new_deps: ValidationDependencies = futures::future::join_all(action_fetches)
+        .await
+        .into_iter()
+        .filter_map(|r| {
+            // Filter out errors, preparing the rest to be put into a HashMap for easy access.
+            match r {
+                (hash, Ok(Some((signed_action, source)))) => {
+                    Some((hash, (signed_action, source).into()))
+                }
+                (hash, Ok(None)) => {
+                    Some((hash, ValidationDependencyState::new(None)))
+                },
+                (hash, Err(e)) => {
+                    tracing::error!(error = ?e, action_hash = ?hash, "Error retrieving prev action");
+                    None
+                }
+            }
+        })
+        .collect();
+
+    current_validation_dependencies.lock().merge(new_deps);
+}
+
+fn get_dependency_hashes_from_actions(actions: impl Iterator<Item = Action>) -> Vec<ActionHash> {
+    actions
+        .flat_map(|action| {
+            vec![
+                match action.prev_action().cloned() {
+                    None => None,
+                    hash => hash,
+                },
+                match action {
+                    Action::Update(action) => Some(action.original_action_address),
+                    Action::Delete(action) => Some(action.deletes_address),
+                    Action::DeleteLink(action) => Some(action.link_add_address),
+                    _ => None,
+                },
+            ]
+            .into_iter()
+            .flatten()
+        })
+        .collect()
+}
+
+/// Examine the list of provided actions and create a list of actions which are sys validation dependencies for those actions.
+/// The actions are merged into `current_validation_dependencies`.
+async fn fetch_previous_actions(
+    current_validation_dependencies: Arc<Mutex<ValidationDependencies>>,
+    cascade: Arc<impl Cascade + Send + Sync>,
+    actions: impl Iterator<Item = Action>,
+) {
+    retrieve_actions(
+        current_validation_dependencies,
+        cascade,
+        get_dependency_hashes_from_actions(actions).into_iter(),
     )
-    .await?;
+    .await;
+}
 
-    Ok(if saturated {
-        WorkComplete::Incomplete(None)
-    } else {
-        WorkComplete::Complete
-    })
+fn get_dependency_hashes_from_ops(ops: impl Iterator<Item = DhtOpHashed>) -> Vec<ActionHash> {
```

Comment:
This function looks good to me, but you could add a test for each op.

Reply:
Yes, not all the op types are covered. This is being called by the validate_op tests and the unit tests but those aren't covering all op types. I think the complex ones are covered so I'd be happy to finish off that test suite as a follow up

Reply:
Ah, yes, it's actually called in the validate op tests. I think that's fine then.